### PR TITLE
Add metric and stat entity

### LINKED ISSUE #53: [FEATURE] A new stats api to save ingestion stats
**Is your feature request related to a problem?**
Return custom geojson import statistics

**What solution would you like?**
A new API that returns statistics about custom geojson import process

**What alternatives have you considered?**
The below API only returns if pipeline is available.
```
curl -X GET "localhost:9200/_nodes/stats/ingest?filter_path=nodes.*.ingest&pretty"
```
Since, API deletes the pipeline once it is completed, this statistics are not available from nodes/stats api.
**Do you have any additional context?**
N/A

### PR DESCRIPTION
**Pull request by VijayanB**
Signed-off-by: Vijayan Balasubramanian <balasvij@amazon.com>

### Description
Following metrics are added for the Upload API

1. id : (string) Unique metric identifier
2. upload: (integer) Number of documents that are requested to be uploaded.
3. duration: (integer) Time, in milliseconds, spent preprocessing during upload.
4. success: (integer) Number of documents that are successfully uploaded.
5. failed: (integer)  Number of documents that are failed to upload.
 
### Issues Resolved
#53 
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/geospatial/action/upload/geojson/Uploader.java [hunk 1]
```diff
@@ -139,4 +150,23 @@ private void indexContentAsDocument(
             uploadStepListener.onFailure(new IllegalStateException(message.toString()));
         }));
     }
+
+    private void injectPostUploadMetrics(BulkResponse response) {
+        int total = response.getItems().length;
+        metricBuilder.uploadCount(total);
+        metricBuilder.duration(response.getTook().duration());
+        if (response.hasFailures()) {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/geospatial/action/upload/geojson/Uploader.java:158]: !hasFailures?

**Review by jmazanec15 — COMMENTED** [review]
Could you define the metrics/stats you are adding in the issue?

File: src/main/java/org/opensearch/geospatial/action/upload/geojson/Uploader.java (hunk 1 above)
Comment (VijayanB) [reply src/main/java/org/opensearch/geospatial/action/upload/geojson/Uploader.java]: Good catch. Added test to validate the same.

File: src/main/java/org/opensearch/geospatial/action/upload/UploadMetric.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    public long getUploadCount() {
+        return uploadCount;
+    }
+
+    /**
+     * @return Metric's Identifier
+     */
+    public String getMetricID() {
+        return metricID;
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/geospatial/action/upload/UploadMetric.java:65]: What is the metricID?

File: src/main/java/org/opensearch/geospatial/action/upload/UploadMetric.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * Builder to create {@link UploadMetric}
+     */
+    public static final class UploadMetricBuilder {
+
+        private long duration;
+        private long failedCount;
+        private String metricID;
+        private long successCount;
+        private long uploadCount;
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/geospatial/action/upload/UploadMetric.java:95]: Do these need to be initialized? Seems like metricID at the least should be a required parameter. What happens if this isnt set?

File: src/main/java/org/opensearch/geospatial/action/upload/UploadStats.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        totalAPICount = new CounterMetric();
+    }
+
+    /**
+     * Add new metric to {@link UploadStats}
+     * @param newMetric
+     */
+    public void addMetric(UploadMetric newMetric) {
+        Objects.requireNonNull(newMetric, "metric cannot be null");
+        metrics.add(newMetric);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/geospatial/action/upload/UploadStats.java:53]: Do we check for duplicate metrics anywhere?

File: src/main/java/org/opensearch/geospatial/action/upload/UploadMetric.java (hunk 2 above)
Comment (VijayanB) [reply src/main/java/org/opensearch/geospatial/action/upload/UploadMetric.java:65]: Unique identifier for every upload request

File: src/main/java/org/opensearch/geospatial/action/upload/UploadMetric.java (hunk 3 above)
Comment (VijayanB) [reply src/main/java/org/opensearch/geospatial/action/upload/UploadMetric.java:110]: Now metric id is required now. Default for all others are 0

File: src/main/java/org/opensearch/geospatial/action/upload/UploadStats.java (hunk 4 above)
Comment (VijayanB) [reply src/main/java/org/opensearch/geospatial/action/upload/UploadStats.java:49]: changed to HashSet and using metric ID to validate duplicate metrics

File: src/main/java/org/opensearch/geospatial/action/upload/geojson/UploadGeoJSONTransportAction.java [hunk 5]
```diff
@@ -47,6 +48,9 @@ public UploadGeoJSONTransportAction(
 
     @Override
     protected void doExecute(Task task, UploadGeoJSONRequest request, ActionListener<UploadGeoJSONResponse> actionListener) {
+        // update metric api count
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/geospatial/action/upload/geojson/UploadGeoJSONTransportAction.java:51]: nit: do we need this comment? Code is pretty straightforward and self explanatory

Comment (VijayanB) [reply src/main/java/org/opensearch/geospatial/action/upload/geojson/UploadGeoJSONTransportAction.java]: Ack

File: src/main/java/org/opensearch/geospatial/action/upload/geojson/Uploader.java [hunk 6]
```diff
@@ -127,16 +132,51 @@ private void indexContentAsDocument(
...
+        UploadMetric metric = createUploadMetric(metricID, response);
+        UploadStats.getInstance().addMetric(metric);
+    }
+
+    private UploadMetric createUploadMetric(String id, BulkResponse response) {
+        UploadMetric.UploadMetricBuilder metricBuilder = new UploadMetric.UploadMetricBuilder(id, GEOJSON);
+        BulkItemResponse[] items = response.getItems();
+        metricBuilder.uploadCount(items.length);
+        metricBuilder.duration(response.getTook().duration());
+        if (!response.hasFailures()) {
```
[the remark this answers is not in the document]
Comment (VijayanB) [reply src/main/java/org/opensearch/geospatial/action/upload/geojson/Uploader.java]: Ack

**Review by martin-gaievski — APPROVED** [review]

**Review by jmazanec15 — APPROVED** [review]

**Pull request merged**